### PR TITLE
More current source

### PR DESCRIPTION
**Pull request by neworderofjamie**
I started switching some of my models over to using current sources and found a few issues:

1. State variables can be initialised on either the GPU or CPU (largely for CPU_ONLY mode/legacy) - in CPU mode, current source variables weren't initialised
2. Current source variables were being read from global memory into registers, but never being written back if they got updated

I fixed both of these as well as updating one of the tests to cover initialising current source variables.

### CODE REVIEW

File: lib/src/standardGeneratedSections.cc [hunk 1]
```diff
@@ -119,33 +118,32 @@ void StandardGeneratedSections::neuronCurrentInjection(
...
-            string iCode = csm->getInjectionCode();
-            substitute(iCode, "$(id)", localID);
-            StandardSubstitutions::currentSourceInjection(iCode, cs,
-                                csVars, csDerivedParams, csExtraGlobalParams,
-                                functions, ftype, rng);
-            os << iCode << std::endl;
+
+        // Read current source variables into registers
+        for(const auto &v : csVars.container) {
+            os <<  v.second << " l" << v.first << " = " << devPrefix << v.first << cs->getName() << "[" << localID << "];" << std::endl;
```
Comment (braintimeException) [inline lib/src/standardGeneratedSections.cc]: In theory there can be several current sources per population and if some of their variables have the same name, there will be a collision. For example, if for any reason I want to inject DC into half of the population, and AC into another half, controlling the injection by, say boolean ``applyCurrent`` variable in both cases, then ``applyCurrent`` will be defined twice. Appending name of the current source solves it.

File: lib/src/standardGeneratedSections.cc [hunk 2]
```diff
@@ -119,33 +118,32 @@ void StandardGeneratedSections::neuronCurrentInjection(
...
+        string iCode = csm->getInjectionCode();
+        substitute(iCode, "$(id)", localID);
+        StandardSubstitutions::currentSourceInjection(iCode, cs,
+                            csVars, csDerivedParams, csExtraGlobalParams,
+                            functions, ftype, rng);
+        os << iCode << std::endl;
+
+        // Write updated variables back to global memory
+        for(const auto &v : csVars.container) {
+             os << devPrefix << v.first << cs->getName() << "[" << localID << "] = l" << v.first << ";" << std::endl;
```
Comment (braintimeException) [inline lib/src/standardGeneratedSections.cc]: Same issue with local variable name as above

File: lib/src/standardSubstitutions.cc [hunk 3]
```diff
@@ -308,7 +308,7 @@ void StandardSubstitutions::currentSourceInjection(
     const std::string &rng)
 {
     substitute(iCode, "$(t)", "t");
-    name_substitutions(iCode, "l", csmVars.nameBegin, csmVars.nameEnd, cs->getName());
+    name_substitutions(iCode, "l", csmVars.nameBegin, csmVars.nameEnd);
```
Comment (braintimeException) [inline lib/src/standardSubstitutions.cc]: see above

File: lib/src/standardGeneratedSections.cc (hunk 1 above)
Comment (neworderofjamie) [reply lib/src/standardGeneratedSections.cc]: This is why I put all the code for each current source in a scope (that's what the ``CodeStream::Scope b(os);`` bit does) - if you look at the generated code it'll be enclosed in {} so the fact the names are the same doesn't matter.

Comment (braintimeException) [reply lib/src/standardGeneratedSections.cc]: ahh my bad! Then this PR is good to go.

Comment (neworderofjamie) [reply lib/src/standardGeneratedSections.cc]: Do you have a fix for the extra global parameters bug we can cherry pick in?

Comment (braintimeException) [reply lib/src/standardGeneratedSections.cc]: I did not commit it yet, so I think I will simply push it here

**Comment by braintimeException** [comment]
I think, once the tests finish, this PR can be merged

**Pull request merged**